### PR TITLE
AVRO-3628: [Java] JUnit 4.x tests are not executed

### PR DESCRIPTION
**Pull request by martin-g**
Use JUnit 5.x Vintage engine instead of JUnit 4.x

## What is the purpose of the change

* Fix the execution of JUnit 4.x unit tests by using JUnit 5.x Vintage engine. Fixes AVRO-3628

## Verifying this change

Check the stdout of the Maven build and make sure that JUnit 4.x tests are executed, e.g. avro-maven-plugin module should execute 11 tests.

## Documentation

- Does this pull request introduce a new feature? no

### CODE REVIEW

File: lang/java/maven-plugin/pom.xml [hunk 1]
```diff
@@ -60,6 +60,7 @@
       <groupId>org.apache.maven</groupId>
       <artifactId>maven-core</artifactId>
       <version>${maven-core.version}</version>
+      <scope>provided</scope>
```
Comment (martin-g) [inline lang/java/maven-plugin/pom.xml:63]: This fixes the following Maven error:
```
[INFO] --- maven-plugin-plugin:3.6.4:helpmojo (generated-helpmojo) @ avro-maven-plugin ---
[ERROR] 

Some dependencies of Maven Plugins are expected to be in provided scope.
Please make sure that dependencies listed below declared in POM
have set '<scope>provided</scope>' as well.

The following dependencies are in wrong scope:
 * org.apache.maven:maven-core:jar:3.3.9:compile
 * org.apache.maven:maven-model:jar:3.3.9:compile
 * org.apache.maven:maven-settings:jar:3.3.9:compile
 * org.apache.maven:maven-settings-builder:jar:3.3.9:compile
 * org.apache.maven:maven-builder-support:jar:3.3.9:compile
 * org.apache.maven:maven-repository-metadata:jar:3.3.9:compile
 * org.apache.maven:maven-artifact:jar:3.3.9:compile
 * org.apache.maven:maven-plugin-api:jar:3.3.9:compile
 * org.apache.maven:maven-model-builder:jar:3.3.9:compile
 * org.apache.maven:maven-aether-provider:jar:3.3.9:compile
```

File: lang/java/maven-plugin/src/test/resources/unit/idl/pom-injecting-velocity-tools.xml [hunk 2]
```diff
@@ -22,7 +22,7 @@
   <parent>
     <artifactId>avro-parent</artifactId>
     <groupId>org.apache.avro</groupId>
-    <version>1.11.0-SNAPSHOT</version>
+    <version>1.12.0-SNAPSHOT</version>
```
Comment (martin-g) [inline lang/java/maven-plugin/src/test/resources/unit/idl/pom-injecting-velocity-tools.xml:25]: It seems something is wrong with the release process. Several pom.xml files were still pointing to the old 1.11.0-SNAPSHOT parent.

File: lang/java/archetypes/avro-service-archetype/src/main/pom/pom.xml [hunk 3]
```diff
@@ -79,9 +79,9 @@
       <version>\${logback.version}</version>
     </dependency>
     <dependency>
-      <groupId>junit</groupId>
-      <artifactId>junit</artifactId>
-      <version>\${junit.version}</version>
+      <groupId>org.junit.vintage</groupId>
+      <artifactId>junit-vintage-engine</artifactId>
+      <version>\${junit5.version}</version>
```
Comment (opwvhk) [inline lang/java/archetypes/avro-service-archetype/src/main/pom/pom.xml:84]: This looks like it (only) uses the vintage engine. Do we also want to run JUnit5 (Jupiter) tests?

In that case, we also need to add the Jupiter engine IIRC.

File: lang/java/maven-plugin/src/test/resources/unit/idl/pom-injecting-velocity-tools.xml (hunk 2 above)
Comment (RyanSkraba) [reply lang/java/maven-plugin/src/test/resources/unit/idl/pom-injecting-velocity-tools.xml:25]: Good catch, this should have been done in https://cwiki.apache.org/confluence/display/AVRO/How+To+Release#HowToRelease-Branching (Step 2c)!

File: lang/java/archetypes/avro-service-archetype/src/main/pom/pom.xml (hunk 3 above)
Comment (RyanSkraba) [reply lang/java/archetypes/avro-service-archetype/src/main/pom/pom.xml:84]: Huh -- as far as I remember @opwvhk is correct, and yet JUnit5 tests [are being run](https://github.com/apache/avro/actions/runs/3044499132/jobs/4904953694#step:7:404)!  (TestDataFile is definitely a JUnit5 migrated class).

Comment (RyanSkraba) [reply lang/java/archetypes/avro-service-archetype/src/main/pom/pom.xml:84]: Oh, my mistake, I see -- yes, this archetype can probably be safely left as it is (on JUnit4) or entirely migrate the generated test to [JUnit5](https://github.com/apache/avro/blob/70260919426f89825ca148f5ee815f3b2cf4764d/lang/java/archetypes/avro-service-archetype/src/main/resources/archetype-resources/src/test/java/integration/SimpleOrderServiceIntegrationTest.java#L33-L35) 

Lets not mix JUnit4 and JUnit5 in the archetype, however!

File: lang/java/maven-plugin/pom.xml [hunk 4]
```diff
@@ -24,7 +24,7 @@
     <artifactId>avro-parent</artifactId>
     <groupId>org.apache.avro</groupId>
     <version>1.12.0-SNAPSHOT</version>
-    <relativePath>../</relativePath>
```
Comment (RyanSkraba) [inline lang/java/maven-plugin/pom.xml:27]: Just for info, is this a best practice?

**Review by RyanSkraba — CHANGES_REQUESTED** [review]

**Comment by RyanSkraba** [comment]
Just for info -- adding up the test cases before the change: **144** versus **10504** after the change :scream:

File: lang/java/archetypes/avro-service-archetype/src/main/pom/pom.xml (hunk 3 above)
Comment (martin-g) [reply lang/java/archetypes/avro-service-archetype/src/main/pom/pom.xml:84]: Good catch! I will migrate the archetype to JUnit 5.x!

File: lang/java/maven-plugin/pom.xml (hunk 4 above)
Comment (martin-g) [reply lang/java/maven-plugin/pom.xml:27]: At least Intellij IDEA suggests so. It shows `../` as a warning.
I'm OK to revert it.

**Comment by martin-g** [comment]
> Just for info -- adding up the test cases before the change: **144** versus **10504** after the change scream

Yup, the CI will become slower again :-)

File: lang/java/archetypes/avro-service-archetype/src/main/pom/pom.xml (hunk 3 above)
Comment (martin-g) [reply lang/java/archetypes/avro-service-archetype/src/main/pom/pom.xml]: Done with https://github.com/apache/avro/pull/1863/commits/65c3ac3354fceaaa6aaa2e81bf54475b1541b2b1

**Review by RyanSkraba — APPROVED** [review]
Thanks for the contribution!

**Pull request merged**